internal/api/sms_provider/huawei_cloud.go

```diff
@@ -0,0 +1,259 @@
+	if request.Body == nil {
+		return []byte(""), nil
+	}
+	bodyByte, err := io.ReadAll(request.Body)
+	if err != nil {
+		return []byte(""), err
+	}
+	request.Body = io.NopCloser(bytes.NewBuffer(bodyByte))
+	return bodyByte, err
+}
+
+// Create a "String to Sign".
+func StringToSign(canonicalRequest string, t time.Time) (string, error) {
+	hashStruct := sha256.New()
+	_, err := hashStruct.Write([]byte(canonicalRequest))
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s\n%s\n%x",
+		SignAlgorithm, t.UTC().Format(DateFormat), hashStruct.Sum(nil)), nil
+}
+
+// Create the HWS Signature.
+func SignStringToSign(stringToSign string, signingKey []byte) (string, error) {
+	hmsha, err := hmacsha256(signingKey, stringToSign)
+	return fmt.Sprintf("%x", hmsha), err
+}
+
+// HexEncodeSHA256Hash returns hexcode of sha256
+func HexEncodeSHA256Hash(body []byte) (string, error) {
+	hashStruct := sha256.New()
+	if len(body) == 0 {
+		body = []byte("")
+	}
+	_, err := hashStruct.Write(body)
+	return fmt.Sprintf("%x", hashStruct.Sum(nil)), err
+}
+
+// Get the finalized value for the "Authorization" header. The signature parameter is the output from SignStringToSign
+func AuthHeaderValue(signatureStr, accessKeyStr string, signedHeaders []string) string {
+	return fmt.Sprintf("%s Access=%s, SignedHeaders=%s, Signature=%s", SignAlgorithm, accessKeyStr, strings.Join(signedHeaders, ";"), signatureStr)
+}
+
+// Signature HWS meta
+type Signer struct {
+	Key    string
+	Secret string
+}
+
+// SignRequest set Authorization header
+func (s *Signer) Sign(request *http.Request) error {
+	var t time.Time
+	var err error
+	var date string
+	if date = request.Header.Get(HeaderXDateTime); date != "" {
+		t, err = time.Parse(DateFormat, date)
+	}
+	if err != nil || date == "" {
+		t = time.Now()
+		request.Header.Set(HeaderXDateTime, t.UTC().Format(DateFormat))
+	}
+	signedHeaders := SignedHeaders(request)
+	canonicalRequest, err := CanonicalRequest(request, signedHeaders)
+	if err != nil {
+		return err
+	}
+	stringToSignStr, err := StringToSign(canonicalRequest, t)
+	if err != nil {
+		return err
+	}
+	signatureStr, err := SignStringToSign(stringToSignStr, []byte(s.Secret))
+	if err != nil {
+		return err
+	}
+	authValueStr := AuthHeaderValue(signatureStr, s.Key, signedHeaders)
+	request.Header.Set(HeaderXAuthorization, authValueStr)
+	return nil
+}
+
+type HuaweiCloudResponse struct {
+	Code        string       `json:"code"`
+	Description string       `json:"description"`
+	Result      []ResultItem `json:"result"`
+}
+
+type ResultItem struct {
+	Total      int    `json:"total"`
+	OriginTo   string `json:"originTo"`
+	CreateTime string `json:"createTime"`
+	From       string `json:"from"`
+	SmsMsgId   string `json:"smsMsgId"`
+	CountryId  string `json:"countryId"`
+	Status     string `json:"status"`
+}
+
+type HuaweiCloudProvider struct {
+	Config *conf.HuaweiCloudProviderConfiguration
+}
+
+// SendMessage implements SmsProvider.
```

Comment:
please indicate the special usage of  `message` in the comment: it should be the SMS Template ID which already registered on huawei cloud.